### PR TITLE
added dexprotector check

### PR DESCRIPTION
**Pull request by Jasi2169**

### CODE REVIEW

File: apkid/rules/apk/packers.yar [hunk 1]
```diff
@@ -27,6 +27,25 @@
 
 include "common.yar"
 
+rule dexprotector : packer
+{
+ /**
+ * DexProtector v6.x.x Both Standard And Business Edition And Demo // Edited By Jasi2169
```
Comment (CalebFenton) [inline apkid/rules/apk/packers.yar]: Could you add https://dexprotector.com/ to this comment? It's nice to know.
Also, you could put "Jasi2169" as the metadata.author if you prefer. No other rules use that yet but if we ever make a feature that uses it, it's a better place to store that info than in a comment.

File: apkid/rules/apk/packers.yar [hunk 2]
```diff
@@ -27,6 +27,25 @@
...
+ **/
+  meta:
+    description = "Dexprotector Licel"
```
Comment (CalebFenton) [inline apkid/rules/apk/packers.yar]: Could you change description to just "DexProtector"? That appears to be the product name.

**Comment by CalebFenton** [comment]
Hey, congrats on being the first PR! Thanks for submitting. Two small suggestions and we should be good to go.

**Comment by Jasi2169** [comment]
well done ,hopefully it is okay now :)
does this open classes.dex in smali ?? Dexguard has its arrays 
anyway i was thinking to code GUI in visual basic .net or java something but not made proper mind
python is not my deal anyway ,hopefully if we get more protectors then more can be added as well maybe more will commit

**Comment by CalebFenton** [comment]
This works by using the [dex module](https://github.com/rednaga/yara/blob/master/libyara/modules/dex.c) we added to yara. It never looks at smali. However, if a packer produces static arrays, those can be used as strings in a yara rule, you just have to know how to look for them with a hex editor. If you tell me the array value in smali and send me a dex file, I can pull the info out. Or you can try your hand at it, but I recommend using 010editor along with https://github.com/strazzere/010Editor-stuff/blob/master/Templates/DEXTemplate.bt.

**Comment by Jasi2169** [comment]
oh cool,well we can get Hex bytes which stays same in obfuscator in every update,because these are also important to patch its anti-tamper checks as far as i have bit of experience in it

as i can see there is in " APKiD/apkid/rules/dex/obfuscators.yar"

  strings:
    $opcodes = {
      00 06 00 01 00 03 00 00 00 00 00 00 00
    
if it does searches bytes like above in classes.dex then there can be chances to add it
PS masks like ?? can be key as well i will see when get time to decompile and get hex bytes from it
Thanks

**Comment by CalebFenton** [comment]
The smali may be the same in every case, but the bytecode will be different because the register mapping and string, method, class indexes will be different. For the rule you mention, I went through and replaced anything that _could_ change between compilations with wildcards. So what remained was basically opcodes (invoke, move, const) and some registers. This is a totally useful approach, it just takes time and it's tedious. You have to basically find a chunk of code with enough parts that don't change that is large enough to be unique. And then you get to lookup each byte against an opcode table to figure out what it does. :)

**Comment by Jasi2169** [comment]
I dont think it is tedous,dont know abt python but When doing with Visual Basic on PC or with java in Android it takes me almost 2min to scan the 40mb classes.dex,  coded some android app before and this actually found to be very useful
when i map for readwrite i get bytes and to patch updates wildcards plays role
example
move result v0  rather than making it 0A 00 i keep 0A ?? cus v0 can be changed in update thats why wildcards will play role here and work on update to catch the packer :)

as far as i know i have seen some obfuscators which keep same algo to decrypt n in updates so byte pattern works great always

here are some example of my apps ex ****2SD 
63 00 ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? 12

***  explorer
12 ?? 71 00 ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? 12

**** SMS
63 ?? ?? ?? 0A ?? 39 ??

now this makes stuff clear which is reliable :)

**Comment by CalebFenton** [comment]
You don't think converting this to Smali isn't tedious? Maybe you are much more patient than me :D

```
00 06 00 01 00 03 00 00 00 00 00 00 00
[20-65]
0c 01 12 12 23 22 ?? ??
1c 03 ?? ?? 12 04 4d 03
02 04 6e 3? ?? ?? 10 02
0c 00 62 01 ?? ?? 12 12
23 22 ?? ?? 12 03 4d 05
02 03 6e 3? ?? ?? 10 02
0c 00 1f 00 ?? ?? 11 00
```

It sounds like you get the idea, but some of your examples look too broad and would likely false positive if scanned against many apks.

**Comment by Jasi2169** [comment]
thats why there should be 2 checks atleast on those cases,as far as i know false positive in my cases will be 1% only cus the api or patterns packers calls is not called in regular exported apk or wither with proguard inbuilt

anyway i did not get this pattern [20-65] ?

**Comment by strazzere** [comment]
`[20-65]` is a [jump](http://yara.readthedocs.io/en/v3.4.0/writingrules.html#hexadecimal-strings) - which is basically a short version of lots of `?` wildcards.

**Comment by CalebFenton** [comment]
Relevant documentation link: https://yara.readthedocs.io/en/v3.5.0/writingrules.html#hexadecimal-strings

**Pull request merged**